File: programmers_level2/타겟_넘버.cpp
#include <vector>

using namespace std;

void DFS(const vector<int> &numbers, const int &target, int &answer, int sum, int cnt) {
    if (cnt == numbers.size()){
        if (sum == target)
            ++answer;
        return; // 모든 숫자를 다 사용하고나면 반드시 return
    }
    DFS(numbers, target, answer, sum+numbers[cnt], cnt+1);
    DFS(numbers, target, answer, sum-numbers[cnt], cnt+1);
}

int solution(vector<int> numbers, int target) {
    int answer = 0;
    DFS(numbers, target, answer, 0, 0);
    return answer;
}

File: programmers/타겟 넘버.py
answer = 0
# DFS 이용
def dfs(idx, result, numbers, target):
    global answer
    if idx == len(numbers): # numbers 원소를 모두 계산했을 경우
        if result == target:
            answer += 1
        return
    else:
        dfs(idx + 1, result + numbers[idx], numbers, target) # 재귀구조
        dfs(idx + 1, result - numbers[idx], numbers, target)

def solution(numbers, target):
    global answer # 전역 변수 사용
    dfs(0, 0, numbers, target) # result = 0으로 초기화
    
    return answer
